docs: improve structure of http example

dario-piotrowicz opened the pull request:
improve the http aio example by:
 - adding a background color to the messages section so that it
   can be distinguished from the others
 - avoid using `hr` elements to divide the various sections, use
   css instead (so that the divisions can also be omitted when
   the sections are not being shown)
 - fix the erroneous presence of an input text element inside a button
   (see: https://github.com/angular/angular/pull/44557#discussion_r787239658)

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## Issue

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
 @jessicajaniuk, @gkalpak :slightly_smiling_face::+1: 
  I didn't end up changing the example too much, hopefully the way in which I updated it is an improvement, please let me know what you think :slightly_smiling_face: 
 
### Divisions
Before:
![divisions-before](https://user-images.githubusercontent.com/61631103/152691431-9119c632-014f-4bac-9733-f71dfa6ccb53.gif)
After:
![divisions-after](https://user-images.githubusercontent.com/61631103/152691441-05536874-4c6c-4d2b-be44-5f62045626c6.gif)

### Edit/Add hero
Before:
![edit-add-hero-before](https://user-images.githubusercontent.com/61631103/152691678-5c5a6be1-a13f-44b0-a25f-031c75a6c274.gif)
After:
![edit-add-hero-after](https://user-images.githubusercontent.com/61631103/152691711-8e1b14d0-f433-4352-9e47-256373ec8820.gif)

dario-piotrowicz committed:
docs: improve structure of http example

improve the http aio example by:
 - adding a background color to the messages section so that it
   can be distinguished from the others
 - avoid using `hr` elements to divide the various sections, use
   css instead (so that the divisions can also be omitted when
   the sections are not being shown)
 - fix the erroneous presence of an input text element inside a button
   (see: https://github.com/angular/angular/pull/44557#discussion_r787239658)

AndrewKushnir reviewed (approved):
Looks great, thanks @dario-piotrowicz 👍

AndrewKushnir commented:
@dario-piotrowicz I noticed that you've mentioned George and Jess in the PR description, so I'm adding them as reviewers as well :)

dario-piotrowicz commented:
@AndrewKushnir thanks a bunch! :smile:
Yeah I mentioned George and Jessica since we started this conversation in a previous PR :slightly_smiling_face::+1:

mary-poppins commented:
You can preview 64197e9 at https://pr44986-64197e9.ngbuilds.io/.
You can preview 57a4176 at https://pr44986-57a4176.ngbuilds.io/.
You can preview 62e8c5b at https://pr44986-62e8c5b.ngbuilds.io/.

jessicajaniuk reviewed (approved):
LGTM :cookie:

Thanks, Dario!

gkalpak reviewed (changes requested):
Thx, @dario-piotrowicz, for going above and beyond (as usual 😛) to improve the docs :100:

TBH, I still find the new behavior a little confusing and unintuitive. I was thinking about what a good and intuitive UI would be and I realized that having the whole list item be a button for the purpose of making the name editable is an common pattern and thus confusing imo.

How about this:
- Instead of having the whole list item be a button, we add a new "edit" button (for example, just before the "delete" button).
- Clicking the "edit" button, would replace the span with the name text with an `<input />`.
- The rest would work the same as it does today (i.e. before this PR).

WDYT?

Another thing to note:
This example is supposed to be modeled after the Tour Of Heroes tutorial examples. So, any changes we make, I think should be made to all toh examples as well. From the HTTP guide:

> The sample app for this guide includes an abridged version of the "Tour of Heroes" example that fetches heroes and enables users to add, delete, and update them. The following sections show examples of the data-update methods from the sample's `HeroesService`.

dario-piotrowicz commented:
Thanks a bunch @gkalpak :slightly_smiling_face:, yeah you do have a point, I'll try to update the edit functionality in the way you mentioned as it does sound more intuitive :slightly_smiling_face::+1: 

___

Regarding the tour of heroes, I'll need to check things out, I thought it was unrelated (as only a single section seems to mention  heroes) :+1: 

___

Regarding this one:
> * The rest would work the same as it does today (i.e. before this PR).

So you don't like my changes? like the removal of the `hr`s and the different style for the messages section? :sob:

gkalpak commented:
Thx, @dario-piotrowicz :+1:

> So you don't like my changes? like the removal of the hrs and the different style for the messages section? 😭

No, no, I was just referring to how the edit functionality works. Your visual changes were fine 👌 :smiley:

dario-piotrowicz commented:
> Thx, @dario-piotrowicz +1
> 
> > So you don't like my changes? like the removal of the hrs and the different style for the messages section? sob
> 
> No, no, I was just referring to how the edit functionality works. Your visual changes were fine ok_hand smiley

ah ok :relieved: 

I'll just update the edit then :slightly_smiling_face:

dario-piotrowicz committed:
fixup! docs: improve structure of http example

mary-poppins commented:
You can preview 1cd7155 at https://pr44986-1cd7155.ngbuilds.io/.

mary-poppins commented:
You can preview b678700 at https://pr44986-b678700.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak please have a look at the updated example :slightly_smiling_face: ([stackblitz link](https://pr44986-b678700.ngbuilds.io/generated/live-examples/http/stackblitz.html))

I've updated it to work pretty much as it did before, but with valid elements (as in, no inputs inside anchors/buttons):
![Peek 2022-02-10 22-41](https://user-images.githubusercontent.com/61631103/153510031-e5fc9238-a95d-498d-a479-585c93695ab7.gif)

Hopefully this is better then the previous version :slightly_smiling_face: 

By the way, the saving happens when you press enter on the input or blur it, I did try different things like toggling the edit button to edit/save, but that started feeling pretty awkward when I added the feature of the input disappearing when you blur the input, I am sure good solution around that could be found, but I felt like is was going a bit overboard, so I figured a simpler version like this could be enough :slightly_smiling_face: 

___

Regarding the alignment with the ToH (the live example here: https://angular.io/tutorial/toh-pt6) it looks to me like the two stackblitz examples are just loosely related and do already work differently (the ToH one for example uses routing for the editing of heroes and has a different search functionality), so I would imagine we should not worry too much about it and my changes are fine on that front (unless we want to actually use this opportunity to align more the two examples) :slightly_smiling_face:

mary-poppins commented:
You can preview 1df1a5b at https://pr44986-1df1a5b.ngbuilds.io/.

gkalpak commented on aio/content/examples/http/src/app/heroes/heroes.component.ts:
```diff
@@ -55,8 +69,16 @@ export class HeroesComponent implements OnInit {
+
+  selectHeroForEditing(hero: Hero) {
     this.editHero = hero;
+    this.changeDetectorRef.detectChanges();
+    this.heroEditInput.get(0)!.nativeElement.value = hero.name;
+    this.heroEditInput.get(0)!.nativeElement.focus();
```
Wat??? We should really not show code like this in the docs :grin:
Could we use property binding to bind value and potentially use the `autofocus` HTML attribute on the `<input>` to have it focused as soon as it is shown?

dario-piotrowicz replied:
@gkalpak of course the binding works nicely, but the `autofocus` unfortunately does not :cry: 

Basically autofocus works the first time I click on an edit button but then stops working, I've stumbled across this post which says that it is the expected behavior: https://netbasal.com/autofocus-that-works-anytime-in-angular-apps-68cb89a3f057

So without autofocus I need to keep the `nativeElement.focus()` in :cry: , also I still need to do ` detectChanges()` so that the element actually appears on the DOM before the `focus()` call :disappointed: 

So I actually just fixed the `nativeElement.value` which wasn't the worst part anyways

What do you think we should do? I am sort of reluctant on creating an autofocus directive just for this... but I guess it'd still be better than the current implementation? (I guess it is sort of a case of deciding which is the lesser evil :sweat:)

We could also think of a different ui for the example, but anything we'd come up with would likely require some autofocus functionality anyways (unless we totally scrap the current idea of the heroes list and go with something completely different)

dario-piotrowicz replied:
I actually went with the setter solution mentioned in the blog post I linked in the previous comment, it looks like a pretty nice implementation actually :slightly_smiling_face: 

what do you think? could this do? :slightly_smiling_face:

gkalpak replied:
This looks good. Thx for looking into it. I think the setter is a nice (and cheap) way out :smiley:

gkalpak commented on aio/content/examples/http/src/app/heroes/heroes.component.ts:
```diff
@@ -24,6 +28,16 @@ export class HeroesComponent implements OnInit {
+      this.add(name);
+    }
+    this.heroName = '';
+    this.editHero = undefined;
```
This line seems redundant.

gkalpak commented on aio/content/examples/http/src/app/heroes/heroes.component.ts:
```diff
@@ -24,6 +28,16 @@ export class HeroesComponent implements OnInit {
       .subscribe(heroes => (this.heroes = heroes));
   }
 
+  addOrEditHero(name: string): void {
```
Is this method used anywhere?

gkalpak commented on aio/content/examples/http/src/app/heroes/heroes.component.css:
```diff
@@ -9,57 +9,71 @@
 .heroes li {
-  display: flex;
-  width: 100%;
+  display: grid;
```
Fancy 😎

dario-piotrowicz replied:
:grin:

gkalpak reviewed (changes requested):
Thx, @dario-piotrowicz :+1:
This looks fine (with a couple minor comments addressed).
Fair enough regarding the ToH example. I am OK with deviating here and definitely won't block this PR for that.

dario-piotrowicz commented:
> Thx, @dario-piotrowicz +1 This looks fine (with a couple minor comments addressed). Fair enough regarding the ToH example. I am OK with deviating here and definitely won't block this PR for that.

Thanks a lot! I did make there a couple of very very silly mistakes! sorry :sob: :bow: 

I'll address your comments :slightly_smiling_face::+1: 

regarding ToH yeah we can continue on that in a follow up PR if want, we can either just distantiate more the two examples (as in also removing the text you cited in which the http guides mentions the ToH example) of try to align them more.

(Personally I don't think there is that much benefit in trying to link the two, if anything it can hinder maintainability :thinking: :man_shrugging: )

dario-piotrowicz committed:
fixup! docs: improve structure of http example

mary-poppins commented:
You can preview 1d32cd5 at https://pr44986-1d32cd5.ngbuilds.io/.

mary-poppins commented:
You can preview a4a290f at https://pr44986-a4a290f.ngbuilds.io/.

gkalpak commented on aio/content/examples/http/src/app/heroes/heroes.component.ts:
```diff
@@ -14,10 +14,14 @@ export class HeroesComponent implements OnInit {
-  @ViewChild('heroNameInput') heroNameInput!: ElementRef;
-  @ViewChildren('heroEditInput') heroEditInput!: QueryList<ElementRef>;
+  @ViewChild('heroEditInput')
+  set input(element: ElementRef<HTMLInputElement>) {
```
Nit: Could you rename this to `heroEditInput` :pray:

dario-piotrowicz replied:
yeah.... I imagine I put `input` just from copy-pasting :sweat_smile:  :+1:

gkalpak commented on aio/content/examples/http/src/app/heroes/heroes.component.ts:
```diff
@@ -76,9 +70,6 @@ export class HeroesComponent implements OnInit {
 
   selectHeroForEditing(hero: Hero) {
```
Now that this method is just `this.editHero = hero` (and it is only used in one place), I think it would make sense to just get rid of it and inline the code.

dario-piotrowicz replied:
sure :slightly_smiling_face::+1:

dario-piotrowicz committed:
fixup! docs: improve structure of http example

mary-poppins commented:
You can preview 2dcc8bd at https://pr44986-2dcc8bd.ngbuilds.io/.

mary-poppins commented:
You can preview 91c5799 at https://pr44986-91c5799.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak please have another look when you can :slightly_smiling_face: :bow: :pray:

gkalpak commented on aio/content/examples/http/src/app/heroes/heroes.component.ts:
```diff
@@ -17,7 +17,7 @@ export class HeroesComponent implements OnInit {
   constructor(private heroesService: HeroesService) {}
 
   @ViewChild('heroEditInput')
```
Could we get rid of the argument now that the property has the same name as the template ref name?

dario-piotrowicz replied:
good spot! :smiley:

dario-piotrowicz replied:
no, seems like the argument is mandatory if I remove it I get an error:
![Screenshot at 2022-02-28 18-47-06](https://user-images.githubusercontent.com/61631103/156040907-75843e6d-27be-4c60-9b5a-391e40fcfda4.png)

are you confusing this with `Input`/`Output`? (I sure did when I saw your comment :smile:)

gkalpak replied:
Doh! Yeah, I confused it with `Input/Output` 🤦‍♂️

dario-piotrowicz replied:
it happens :stuck_out_tongue::+1:

gkalpak commented on aio/content/examples/http/src/app/heroes/heroes.component.ts:
```diff
@@ -1,4 +1,4 @@
-import { Component, OnInit } from '@angular/core';
+import { ChangeDetectorRef, Component, ElementRef, OnInit, QueryList, ViewChild, ViewChildren } from '@angular/core';
```
Some of these imports are no longer needed:

```suggestion
import { Component, ElementRef, OnInit, ViewChild } from '@angular/core';
```

dario-piotrowicz replied:
:cry: :bow:

dario-piotrowicz replied:
sorry :sob:

dario-piotrowicz committed:
fixup! docs: improve structure of http example

mary-poppins commented:
You can preview 1cf1e06 at https://pr44986-1cf1e06.ngbuilds.io/.

gkalpak reviewed (approved):
LGTM :tada:

:shipit:

jessicajaniuk commented:
This PR was merged into the repository by commit 9366a3c5f39853321c5871b8425989b0216a0053.
